pkg/aws/eni/node.go

```diff
@@ -528,6 +536,23 @@ func (n *Node) ResyncInterfacesAndIPs(ctx context.Context, scopedLog *logrus.Ent
 				return nil
 			}
 
+			// The limits include the primary IP, so we need to take it into account
+			// when computing the effective number of available addresses on the ENI.
+			effectiveLimits := limits.IPv4 - 1
+
+			// Include the primary IP when UsePrimaryAddress is set to true on ENI spec.
+			if n.k8sObj.Spec.ENI.UsePrimaryAddress != nil && *n.k8sObj.Spec.ENI.UsePrimaryAddress {
+				effectiveLimits++
+			}
+			if n.node.Ops().IsPrefixDelegated() {
+				effectiveLimits = effectiveLimits * option.ENIPDBlockSizeIPv4
+			}
+
+			availableOnENI := math.IntMax(effectiveLimits-len(e.Addresses), 0)
+			if availableOnENI > 0 {
+				remainAvailableENIsCount++
+			}
+
```

Comment:
This block is copied over directly from `PrepareIPAllocation`, correct? Can we move that logic into a separate function somehow? Otherwise we run into the risk that if one place is changed, that the other one will not, and thus lead to inconsistencies.

Reply:
resloved in https://github.com/cilium/cilium/pull/20755/files#diff-30611d880237aa1610d4ba922a8639e304d4f1e398b98418bdd4200ee4536066R222